### PR TITLE
[no-Jira] Improve test coverage

### PR DESCRIPTION
**Pull request by canac**
Increase the test coverage of some of the most frequently-modified files.

### CODE REVIEW

**Review by wrandall22 — CHANGES_REQUESTED** [review]
For the most part this looks really good and adds some good code coverage, thanks!

Some things to note:
1) You asked about commits, some things you could have done, which may have been slightly nicer, would have been to put the commit "Remove redundant GET request" after the test commits (TDD principle). You could also have combined the cleanup commits for `eventForm.js` into a single commit and have the description be a bulleted list of the cleanup actions done.

2) There are a few places where the test sentencing is a bit weird, I believe you did this to avoid putting a `describe` block around a single `it` block but retain the function name. There are probably trade-offs to readability and useless boilerplate here. I'm ok either way on it, but it does seem a tad jarring to read.

3) There are quite a few places (new and old) where tests are referencing hard-coded UUIDs and text. We should really work towards removing this habit from the codebase and use the reference locations of these values instead. Otherwise, if we change a value in the `testData.spec.js` file we'll have to search and replace it all over the codebase instead of it just working properly.

File: test/spec/controllers/eventForm.spec.js [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+
+      expect(modalMessage.error).toHaveBeenCalledTimes(1);
+      const errorMessage = modalMessage.error.calls.argsFor(0)[0].message;
+
+      expect(errorMessage).toContain('<li>Question</li>');
+    });
+  });
+
+  it('addNewPage adds a new page', () => {
+    scope.addNewPage();
```
Comment (dr-bizz) [inline test/spec/controllers/eventForm.spec.js:290]: Should you add a expect the pages to have a legnth of 2 first, so we know a page was added.

File: test/spec/controllers/eventRegistrations.spec.js [hunk 2]
```diff
@@ -25,37 +23,236 @@ describe('Controller: eventRegistrations', function () {
...
+  });
+
+  it('toggleColumn toggles the column visibility', () => {
+    const setItem = spyOn($window.localStorage, 'setItem');
+    scope.toggleColumn(0);
+
+    expect(scope.blocks[0].visible).toBe(true);
+    expect(setItem).toHaveBeenCalledWith(
+      `visibleBlocks:${testData.conference.id}`,
+      '["e088fefc-eb9c-4904-b849-017facc9e063"]',
```
[the remark this answers is not in the document]
Comment (dr-bizz) [reply test/spec/controllers/eventRegistrations.spec.js:141]: Maybe make the uuid into a const which we export from testData.spec.js?

File: test/spec/controllers/eventRegistrations.spec.js [hunk 3]
```diff
@@ -25,37 +23,236 @@ describe('Controller: eventRegistrations', function () {
...
+        Group: true,
+        GroupId: false,
+        Started: true,
+        Completed: true,
+      });
+    });
+  });
+
+  it('toggleColumn toggles the column visibility', () => {
+    const setItem = spyOn($window.localStorage, 'setItem');
```
Comment (dr-bizz) [inline test/spec/controllers/eventRegistrations.spec.js:135]: Would it be worth testing multiple columns being toggled?

File: test/spec/controllers/eventRegistrations.spec.js [hunk 4]
```diff
@@ -25,37 +23,236 @@ describe('Controller: eventRegistrations', function () {
...
+    });
+
+    it('returns "" if the answer does not exist', () => {
+      scope.queryParameters.orderBy = 'last_name';
+
+      expect(scope.answerSort(testData.registration.registrants[0])).toBe('');
+    });
+
+    it('returns the value of simple answers', () => {
+      scope.queryParameters.orderBy = '9b83eebd-b064-4edf-92d0-7982a330272a';
```
Comment (dr-bizz) [inline test/spec/controllers/eventRegistrations.spec.js:217]: As Bill was mentioning, we should prob make this into a variable, so it's easier to read and not mess up the test.

File: test/spec/controllers/eventRegistrations.spec.js [hunk 5]
```diff
@@ -25,37 +23,236 @@ describe('Controller: eventRegistrations', function () {
...
+
+      expect(scope.answerSort(testData.registration.registrants[0])).toBe(
+        'Test,Person',
+      );
+    });
+
+    it('returns the keys of checkbox answers', () => {
+      scope.queryParameters.orderBy = '18ccfb09-3006-4981-ab5e-405ccf2aad1c';
+
+      expect(scope.answerSort(testData.registration.registrants[0])).toBe(
```
Comment (dr-bizz) [inline test/spec/controllers/eventRegistrations.spec.js:233]: Should it return either it was true or false as well?

Comment (dr-bizz) [reply test/spec/controllers/eventRegistrations.spec.js:233]: I guess it doesn't need to as ir's just ordering the answers

File: test/spec/controllers/eventRegistrations.spec.js [hunk 6]
```diff
@@ -150,4 +348,223 @@ describe('Controller: eventRegistrations', function () {
...
+
+      $httpBackend.flush();
+
+      expect(registrant.checkedInTimestamp).toBeNull();
+    });
+  });
+
+  describe('deleteRegistrant', () => {
+    it('deletes a registrant in a group', () => {
+      $httpBackend.expectDELETE(/^registrants\/.+$/).respond(204, '');
```
Comment (dr-bizz) [inline test/spec/controllers/eventRegistrations.spec.js:536]: Can we do a test where there are more than just one registrant in a group when one is deleted? This is linked to my comment about the error when deleting a registrant in the modal.

Comment (dr-bizz) [reply test/spec/controllers/eventRegistrations.spec.js:536]: Maybe this is doing that. But the error I had is around the same code

File: test/spec/controllers/reviewRegistration.spec.js [hunk 7]
```diff
@@ -40,6 +186,50 @@ describe('Controller: ReviewRegistrationCtrl', function () {
...
+      it('is false on completed registrations', () => {
+        testData.conference.registrantTypes[1].acceptPayOnSite = true;
+        testData.registration.completed = true;
+        initController();
+
+        expect(scope.acceptedPaymentMethods().acceptPayOnSite).toBe(false);
+      });
+    });
+
+    it('returns false when no payment methods are accepted', () => {
```
Comment (dr-bizz) [inline test/spec/controllers/reviewRegistration.spec.js:218]: Is this the test when the form requires no payment?

**Review by dr-bizz — CHANGES_REQUESTED** [review]

File: test/spec/controllers/reviewRegistration.spec.js (hunk 7 above)
Comment (canac) [reply test/spec/controllers/reviewRegistration.spec.js:226]: Yes. Is there anything I can do to make that clearer?

File: app/scripts/controllers/eventRegistrations.js [hunk 8]
```diff
@@ -573,43 +573,41 @@ angular
...
-            $http
-              .get('registrations/' + registrant.registrationId)
-              .then(function (response) {
-                var registration = response.data;
-                var url = 'registrations/' + registration.id;
-
-                if (registration.registrants.length > 1) {
-                  //Delete Registrant
-                  url = 'registrants/' + registrant.id;
+            const registration = $scope.getRegistration(
```
[the remark this answers is not in the document]
Comment (canac) [reply app/scripts/controllers/eventRegistrations.js]: Upon further testing, removing that GET request seemed to cause issues, so I reverted that change.

File: test/spec/controllers/eventRegistrations.spec.js (hunk 6 above)
Comment (canac) [reply test/spec/controllers/eventRegistrations.spec.js:566]: Yes, this tests removing one registrant from a group with multiple registrants. I reverted the change to the controller, and that should fix the error.

File: app/scripts/controllers/eventForm.js [hunk 9]
```diff
@@ -210,28 +208,29 @@ angular
           });
       };
 
-      var makePositionArray = function () {
-        var tempPositionArray = [];
+      // Generate a map of blockIds to their page and index within the page
+      var makeBlockPositionMap = function () {
+        var tempPositionArray = {};
```
[the remark this answers is not in the document]
Comment (canac) [reply app/scripts/controllers/eventForm.js]: I fixed the variable name

File: test/spec/controllers/reviewRegistration.spec.js (hunk 7 above)
Comment (dr-bizz) [reply test/spec/controllers/reviewRegistration.spec.js:221]: No you're good, I think I was just confirming

**Review by dr-bizz — APPROVED** [review]
Thank you for adding this coverage!

**Comment by canac** [comment]
@wrandall22 Do you want to take another look, or should I dismiss your review?

**Pull request merged**